jax_omeroutils/xml_editor.py

```diff
@@ -0,0 +1,234 @@
+        proj_count = 1
+        ds_count = 1
+        for project in proj_ds.keys():
+            id = f"Project:{proj_count}"
+            proj_count += 1
+            name = project
+            proj = create_proj(id=id, name=name)
+            for dataset in proj_ds[project]:
+                id = f"Dataset:{ds_count}"
+                ds_count += 1
+                ds, ds_ref = create_dataset_and_ref(id=id, name=dataset)
+                newome.datasets.append(ds)
+                proj.dataset_ref.append(ds_ref)
+            newome.projects.append(proj)
+    return newome
+
+
+def add_screens(ome, imp_json):
+    newome = copy.deepcopy(ome)
+    columns = imp_json['user_supplied_md']['file_metadata'][0].keys()
+    if 'screen' not in columns:
+        return ome
+    else:
+        md = imp_json['user_supplied_md']['file_metadata']
+        scr_count = 1
+        screens = [i['screen'] for i in md]
+        for scr in list(set(screens)):
+            id = f"Screen:{scr_count}"
+            scr_count += 1
+            scr_obj = create_screen(id=id, name=scr)
+            newome.screens.append(scr_obj)
+    return newome
+
+
+def add_annotations(ome, imp_json):
+    columns = list(imp_json['user_supplied_md']['file_metadata'][0].keys())
+    if 'project' in columns:
+        newome = add_annotations_images(ome, imp_json)
+        return newome
+    elif 'screen' in columns:
+        newome = add_annotations_plates(ome, imp_json)
+        return newome
+    return ome
+
+
+def add_annotations_images(ome, imp_json):
+    newome = copy.deepcopy(ome)
+    columns = list(imp_json['user_supplied_md']['file_metadata'][0].keys())
+    if 'project' in columns:
+        columns.remove('project')
+    if 'dataset' in columns:
+        columns.remove('dataset')
+    md = imp_json['user_supplied_md']['file_metadata']
+    max_ann = 0
+    for ann in newome.structured_annotations:
+        clean_id = int(ann.id.split(":")[-1])
+        if clean_id > max_ann:
+            max_ann = clean_id
+    ann_count = max_ann + 1
+    for line in md:
+        filename = line['filename']
+        ann_dict = {i: line[i] for i in columns}
+        ann_dict.pop('filename')
+        ann_dict = {k: v for k, v in ann_dict.items() if isinstance(v, str)}
+        for ann in ome.structured_annotations:
+            if isinstance(ann, CommentAnnotation):
+                src_file = ann.value.split(sep)[-1]
+                if src_file == filename:
+                    img_id = ann.namespace
+                    for i in newome.images:
+                        if i.id == img_id:
+                            mmap = []
+                            for _key, _value in ann_dict.items():
+                                if _value:
+                                    mmap.append(M(k=_key, value=str(_value)))
+                                else:
+                                    mmap.append(M(k=_key, value=''))
+                            ann, annref = create_kv_and_ref(
```

Comment:
Would be easier to read this for loop if the `ann` variable wasn't overwritten. Could call this one `newann` or call the loop variable `filename_ann` to make it more clear we're looping through to find filenames in the CommentAnnotations